### DISCUSSION TITLE
Including subclasses in type annotations for "repositories" declared types

### DISCUSSION THREAD

**Discussion by diogo-rossi** [haiiliin/abqpy#2782] [Ideas]
This idea came out when dealing with issues of type hints in #2758, but is not directly related.

## Problem

One issue often appears when Abaqus stores new created objects in "repositories", which are dictionaries mapping names to superclasses objects. The issue is the inability to access methods of the subclasses.

#### Example:
the `mdb.jobs` dictionary

https://github.com/haiiliin/abqpy/blob/907a502aa58aa2b6689263e1ec844835114b2d3e/src/abaqus/Mdb/MdbBase.py#L32-L33

When creating a new Job from a specific method, the subclass methods are not accessible 
```python
mdb.Job(name='Job-1', ...) # creates a 'ModelJob'
mdb.jobs['Job-1'].writeInput(consistencyChecking=OFF) # The method is not accessible by the language server
```
The method `writeInput` is not accessible by the language server because the dictionary is of type `Dict[str, Job]` and `Job` object does not have the method `writeInput`.

![image](https://user-images.githubusercontent.com/57204023/195945723-49cc2a95-c888-4cfc-b18f-00879594f467.png)

## Possible solution:

Add all possible subclasses in the dictionary declaration.

#### Example:
In the `mdb.jobs` dictionary
```python
 #: A repository of Job objects. 
 jobs: Dict[str, Union[Job, ModelJob, JobFromInputFile]] = {}
```

**Comment by haiiliin** [comment]
Right, this is what it should be
